### ISSUE TITLE
Add module descriptor

### ISSUE THREAD

**Issue by XakepSDK** [typetools/checker-framework#4559]
Add module descriptor (`module-info.java`)
It's possible to use multi-release jar to support both Java 8 and have module descriptor
Compiled module descriptor should be here `META-INF/versions/9/module-info.class`

**Comment by mtf90** [comment]
I am currently in the process of modularizing some projects in order to use `jlink` (and later `jpackage`) for distributing software and came across the issue that checkerframework only provides module support via the `Automatic-Module-Name: org.checkerframework.checker.qual` entry in the `MANIFEST.MF`. This causes problems with `jlink`:

```
[ERROR] Error: automatic module cannot be used with jlink: org.checkerframework.checker.qual from file:///home/frohme/.m2/repository/org/checkerframework/checker-qual/3.40.0/checker-qual-3.40.0.jar
```

Are there any plans for (or is there at least any interest in) adding proper modulization support to checkerframework (at least for the client-facing artifacts such as `checker-qual`)?

Since I am in a similar boat of wanting to add modulization support without breaking Java 8 compatibility, I can report that this can be relatively cleanly implemented with Multi-Release JARs indeed. There are only small "drawbacks" to this approach:

* You now need a Java 9+ JDK to build checkerframework
  * Between Java 9-11 there is [some wonkiness](https://bugs.openjdk.org/browse/JDK-8211194) in the Javadoc generation so that it may happen that people on Java 8 cannot properly reference, e.g., JDK11 generated Javadoc (I had a similar issue with recent guava releases). However, given that you seem to build on JDK17 anyway (see the links to `java.lang` [here](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/Nullable.html)), you are already affected by this problem or the fix thereof.
* The build process gets a little bit bloated since you would now need three compilation steps.

If wanted, I can try to sketch an initial PR draft for this feature. However, since I am more used to Maven than Gradle and I am rather unfarmiliar with all the intrinsics of the checkerframework build process, I'll probably need some hand holding later.

**Comment by mernst** [comment]
@mtf90 Thanks for bringing this up.  Modularization isn't high on our priority list, but we would welcome it.  If you can take the lead, we would gladly help.  Thank you!

Requiring a Java 9+ JDK to build the Checker Framework is not a problem.

**Comment by mtf90** [comment]
Just a heads-up for everyone that faces similar `jlink`-/`jpackage`-related problems as mentioned in https://github.com/typetools/checker-framework/issues/4559#issuecomment-1830070216: A potential workaround to this issue is to declare the dependency on `org.checkerframework.checker.qual` as `transitive` so that it won't be included in the `jlink`ed/`jpackage`d bundle. This should not cause any problems, as [JLS 13.5.8](https://docs.oracle.com/javase/specs/jls/se21/html/jls-13.html#jls-13.5.8) states that

> Adding or deleting annotations has no effect on the correct linkage of the binary representations of programs in the Java programming language.

In case of a distributed application the missing annotations should not matter, because the user of the application typically does not care about the application's internals (e.g., which classes it uses, etc.). In case of a library-like artifact, any code that would make use of the annotations (e.g., for nullness analysis) would add its own dependency on `checker-qual` making the annotations available again when loading (and analyzing) the code even with the `transitive` dependency declaration.

This should take some pressure off this issue, as I mentioned in https://github.com/typetools/checker-framework/pull/6326#issuecomment-1895450887 that it may take some effort to support modules cleanly. For now, an `Automatic-Module-Name` may be the best we can get.

**Comment by sgammon** [comment]
#6326 was merged; I think this can be closed @mernst 😄

**Comment by mernst** [comment]
@sgammon Thanks for the ping.

**Comment by mtf90** [comment]
Just a heads-up: While these changes now nicely allow for jlink-ing applications that use checkerframework, running the javadoc generation now raises the following warning:

```
javadoc: warning - The code being documented uses modules but the packages defined in https://checkerframework.org/releases/3.48.3/api/ are in the unnamed module.
```

(this started with 3.43.0 which included this change).


This is not a critical problem, as the generated documentation still references the annotations just fine. I just wanted to notify you (and others) about this and ask whether it would make sense to open a separate issue for tracking this when bumping the minimum language version to Java 11?

**Comment by mtf90** [comment]
@cowwoc IIRC, the problem was that the qualifier documentation referenced code from the main checker implementation, so we relied on gradle-magic to build one big classpath where everything could be resolved. Especially, generation happens in "java 8 mode" which is why the javadoc tool probably doesn't generate any module information (`module-info.java` is being completely ignored by the tool).

Having a consistent experience would be nice but I'm afraid that it would require quite some refactoring, which is why I proposed to put it on the "Java 11" list for the future.
